pyriemann/utils/kernel.py

```diff
@@ -0,0 +1,97 @@
+import numpy as np
+
+from .mean import mean_riemann
+from .base import invsqrtm, logm
+
+
+def kernel(X, Y=None, Cref=None, metric='riemann'):
+    r""" Calculates the Kernel matrix K of inner products of two sets
+         X and Y of SPD matrices on tangent space of C according to the
+         specified metric.
+
+        Parameters
+        ----------
+        X : ndarray, shape (n_matrices, n_channels, n_channels)
+            First set of SPD matrices.
+        Y : ndarray, shape (n_matrices, n_channels, n_channels), optional
+            Second set of SPD matrices. If None, Y is set to X.
+        Cref : ndarray, shape (n_channels, n_channels), optional
+            Reference point for the tangent space and inner product
+            calculation. If None, Cref is calculated as the mean of X
+            according to the specified metric.
+        metric : string (default: 'riemann')
+            The type of metric used for tangent space and mean estimation. Can
+            be 'riemann'.
+
+        Returns
+        ----------
+        K : ndarray, shape (n_matrices, n_matrices)
+            The kernel matrix of X and Y.
+
+        Notes
+        -----
+        .. versionadded:: 0.2.8
+        """
+    if metric == 'riemann':
+        return kernel_riemann(X, Y, Cref)
+    else:
+        raise ValueError("Kernel metric must be 'riemann'.")
+
+
+def kernel_riemann(X, Y=None, Cref=None):
+    r""" Calculates the Kernel matrix K of inner products of two sets
+     X and Y of SPD matrices on tangent space of C by calculating pairwise
+
+    .. math::
+        K_{i,j} =
+        {tr}(\log(C^{-1/2}X_i C^{-1/2})\log(C^{-1/2}Y_j C^{-1/2}))
+
+    Parameters
+    ----------
+    X : ndarray, shape (n_matrices, n_channels, n_channels)
+        First set of SPD matrices.
+    Y : ndarray, shape (n_matrices, n_channels, n_channels), optional
+        Second set of SPD matrices. If None, Y is set to X.
+    Cref : ndarray, shape (n_channels, n_channels), optional
+        Reference point for the tangent space and inner product calculation.
+        If None, Cref is calculated as the Riemannian mean of X.
+
+    Returns
+    ----------
+    K : ndarray, shape (n_matrices, n_matrices)
+        The kernel matrix of X and Y.
+
+    Notes
+    -----
+    .. versionadded:: 0.2.8
+    """
+    if Cref is None:
+        G = mean_riemann(X)
+        G_invsq = invsqrtm(G)
+
+    else:
+        G_invsq = invsqrtm(Cref)
+
+    n_matrices_X, n_channels, n_channels = X.shape
+
+    X_ = np.matmul(G_invsq, np.matmul(X, G_invsq))
```

Comment:
I don't think that works, because X has shape (n_matrices, n_channels, n_channels) and G_invsq has shape(n_channels, n_channels). It needs to be multiplied from both sides to each matrix in X and matmul can do this vectorized (and thus in parallel).